lib/libutee/arch/arm/ftrace/ftrace.c

```diff
@@ -0,0 +1,127 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2019, Linaro Limited
+ */
+
+#include <string.h>
+#include <user_ta_header.h>
+#include <utee_syscalls.h>
+
+#define FTRACE_RETFUNC_DEPTH		50
+
+static unsigned long ftrace_ret_stack[FTRACE_RETFUNC_DEPTH];
+static unsigned int ftrace_ret_idx;
+static const char hex_str[] = "0123456789abcdef";
+
+void __ftrace_return(void);
+void ftrace_enter(unsigned long pc, unsigned long *lr);
+
+static void __noprof ftrace_memmove(struct ftrace_buf *fbuf, size_t size)
+{
+	char *dst = (char *)fbuf + fbuf->buf_off;
+	const char *src = (char *)fbuf + fbuf->buf_off + size;
+	size_t len = 0;
+
+	fbuf->curr_size -= size;
+	len = fbuf->curr_size;
+
+	while (len--)
```

Comment:
suggest `memcpy(dst, src, len)` or `TEE_MemMove(dst, src, len)`.

Reply:
I couldn't use any library apis inside this __noprof ftrace function as those apis could be instrumented and would create a loop.